# rules for alignment
#
# require config: 
#   DATA_DIR:
#   GENOMES: [geno1,geno2]
#   star_hgmm: extra
#   READ_LENGTH
# todo: hg38mm10 to genome=hg38.mm10, followed by extract
from os import path

rule combine_genome:
    input:
        fa1=f"{DATA_DIR}/{GENOMES[0]}/genome.fa",
        fa2=f"{DATA_DIR}/{GENOMES[1]}/genome.fa",
        gtf1=f"{DATA_DIR}/{GENOMES[0]}/genes.gtf",
        gtf2=f"{DATA_DIR}/{GENOMES[1]}/genes.gtf",
    output:
        fa=f"{DATA_DIR}/{GENOMES[0]}{GENOMES[1]}/genome.fa",
        gtf=f"{DATA_DIR}/{GENOMES[0]}{GENOMES[1]}/genes.gtf",
    shell: """
    """

rule seperate_genome:
    input: {prefix}.bam
    output:
        bam="{prefix}.{genome}.bam"
    threads: 10
    params:
        extra="-q10"
    shell: """
        genoChr=${{genome}}Chr
        tmpDir=$(mktemp -d)
        samtools view -H {input} \
            | sed -n "/^@SQ/! p; /SN:$genoChr/ p" \
            | sed "s/SN:$genoChr/SN:chr/" > $tmpDir/header.sam
        samtools view {input} {params.extra} -@ {threads} \
            | sed -n "/\t$genoChr/ p" \
            | sed "s/$genoChr/chr/" > $tmpDir/body.sam
        cat $tmpDir/header.sam $tmpDir/body.sam | samtools view -Shb - > {output.bam}
        rm -r $tmpDir
        """
